[TensorPipe] Do not mark future messages as complete after they have timed out

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39012 [RPC] Fix flaky test by waiting for async rref calls
* #39011 [TensorPipe] Detect duplicate worker names
* #39010 [TensorPipe] Improve serialization
* #38934 [TensorPipe] Close and join TP context at shutdown
* #38933 [TensorPipe] Implement join correctly
* #38930 [TensorPipe] Always complete futures from thread pool
* **#38931 [TensorPipe] Do not mark future messages as complete after they have timed out**
* #38929 [TensorPipe] Fix lock inversion upon response read error handling
* #38928 [TensorPipe] Fix timeout computation
* #38927 [TensorPipe] Add cases for TP in RPC test helpers
* #38926 [TensorPipe] Pass names of endpoints to context/pipe for easier debugging



When requests time out they are not aborted, so they could in fact still complete successfully but, when they do so, they try to mark an errored future as complete, which causes an error. I don't see any atomic way of doing future->markCompleteIfNeeded, so we implement it on top of it on our side.

Differential Revision: [D21703015](https://our.internmc.facebook.com/intern/diff/D21703015/)

Differential Revision: [D21703015](https://our.internmc.facebook.com/intern/diff/D21703015)

lw committed:
[TensorPipe] Do not mark future messages as complete after they have timed out

When requests time out they are not aborted, so they could in fact still complete successfully but, when they do so, they try to mark an errored future as complete, which causes an error. I don't see any atomic way of doing future->markCompleteIfNeeded, so we implement it on top of it on our side.

Differential Revision: [D21703015](https://our.internmc.facebook.com/intern/diff/D21703015/)

[ghstack-poisoned]

lw committed:
Update on "[TensorPipe] Do not mark future messages as complete after they have timed out"




When requests time out they are not aborted, so they could in fact still complete successfully but, when they do so, they try to mark an errored future as complete, which causes an error. I don't see any atomic way of doing future->markCompleteIfNeeded, so we implement it on top of it on our side.

Differential Revision: [D21703015](https://our.internmc.facebook.com/intern/diff/D21703015/)

Differential Revision: [D21703015](https://our.internmc.facebook.com/intern/diff/D21703015)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
Lint error is real:

```
[
  {
    path: 'torch/csrc/distributed/rpc/tensorpipe_agent.cpp',
    start_line: 670,
    end_line: 670,
    start_column: 42,
    end_column: 42,
    annotation_level: 'failure',
    message: "[performance-unnecessary-value-param] warning: the parameter 'futureMessage' is copied for each invocation but only used as a const reference; consider making it a const reference"
  },
  {
    path: 'torch/csrc/distributed/rpc/tensorpipe_agent.cpp',
    start_line: 678,
    end_line: 678,
    start_column: 42,
    end_column: 42,
    annotation_level: 'failure',
    message: "[performance-unnecessary-value-param] warning: the parameter 'futureMessage' is copied for each invocation but only used as a const reference; consider making it a const reference"
  }
]
```

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -131,6 +131,17 @@ class TensorPipeAgent : public RpcAgent {
+  // only if it isn't yet. It does exist for errors (setErrorIfNeeded) but, even
+  // then, it ends up printing a log message, which may worry the user. To solve
+  // both issues we use a separate atomic flag to know the status of the future.
+  struct AtomicFutureMessage {
```
Should we instead add a "test-and-set" API into the Future type? I would assume this is also a problem for ProcessGroup agent?

mrshenli replied:
As there are many PRs depends on this one and as this is localized change, I think we can get this one in once the lint error is gone. Could you please create an issue for adding "test-and-set" API to make sure we don't forget about it?

osalpekar replied:
@mrshenli @lw I tried implementing `markCompletedIfNeeded` similar to `setErrorIfNeeded` here: https://github.com/pytorch/pytorch/pull/39080. Maybe this API change could simplify this PR?

rohan-varma replied:
FWIW, I don't think this is a problem for the PG agent as we have a `futures_` map and we remove from that map if the future times out. In that case, we don't mark the future as completed a 2nd time because it's not found in the map .

osalpekar replied:
I just added a comment on https://github.com/pytorch/pytorch/issues/39082 - I think we actually may not necessarily need this change if we modify the data structure for storing timeout candidates. We can still keep the "test-and-set" API as a generic utility.

lw replied:
Another problem with `setErrorIfNeeded` is that it always prints a log message which warns that it had to discard an error. Currently, the RPC thread will eventually time out all futures, which means the above warning would be printed for all messages, which is very noisy. If we could find a way to suppress those warnings then we could definitely go for this fix.

In #39082 other solutions are proposed and I'll comment there. I'd be inclined to land this as it is and then fix it properly later, so I don't have to resolve conflicts in all the diffs above this one in the stack ;)

lw commented:
I had seen the lint errors but I didn't fix them because I actually disagree with them: it's true that `shared_ptr` is technically a class and that by declaring that argument by value we may end up making a copy, but if one thinks of `shared_ptr` as a glorified pointer, then it looks really weird to take a reference to a pointer. However, this made me realize that in fact we could just take a reference to the WrappedFutureMessage, removing the `shared_ptr` from the signature entirely. Hopefully this makes the linter happy.

(Note that in the following commit in the stack we'll actually need access to the shared_ptr by value because we'll be moving it into the closure of a lambda, so we'll have to re-introduce the shared_ptr... I guess all the discussion above was purely academic ;) )

As for your other comments, I think the ProcessGroup agent doesn't need this because it removed the future from its internal state once it times out, so there is only one copy of it at all times. But I agree that what we're doing here is rather ugly so I opened #39082 to remember to fix it somehow.

lw committed:
Update on "[TensorPipe] Do not mark future messages as complete after they have timed out"




When requests time out they are not aborted, so they could in fact still complete successfully but, when they do so, they try to mark an errored future as complete, which causes an error. I don't see any atomic way of doing future->markCompleteIfNeeded, so we implement it on top of it on our side.

Differential Revision: [D21703015](https://our.internmc.facebook.com/intern/diff/D21703015/)

Differential Revision: [D21703015](https://our.internmc.facebook.com/intern/diff/D21703015)

[ghstack-poisoned]

mrshenli reviewed (approved):
Test failures are irrelevant.
